api/mapping/services_nlp.py

```diff
@@ -1,19 +1,305 @@
+    scan_report_id = field.scan_report_table.scan_report.id
+
+    # Define NLP things
+    url = "https://ccnett2.cognitiveservices.azure.com/text/analytics/v3.1-preview.3/entities/health/jobs?stringIndexType=TextElements_v8"
+    headers = {
+        "Ocp-Apim-Subscription-Key": os.environ.get("NLP_API_KEY"),
+        "Content-Type": "application/json; utf-8",
+    }
+
+    # Create empty list to later hold job URLs
+    post_response_url = []
 
+    # Checks to see if the field is 'pass_from_source'
+    # If True, we pass field-level data i.e. a single string (field description)
+    # If False, we pass all values associated with that field
+    if field.pass_from_source:
+
+        print(">>> Working at field level...")
+
+        # We want to use the field description if available
+        # However, we fall back to field name if field_description is None
+        if field.field_description is None:
+            document = {
+                "documents": [
+                    {"language": "en", "id": field.id,
+                        "text": field.name.replace("_", " ")}
+                ]
+            }
+
+        else:
+            # Create a single dictionary item for the field description
+            # Convert to JSON for NLP, POST the data to NLP API, save the job URL
+            document = {
+                "documents": [
+                    {"language": "en", "id": field.id,
+                        "text": field.field_description.replace("_", " ")}
+                ]
+            }
+
+        payload = json.dumps(document)
+        response = requests.post(url, headers=headers, data=payload)
+        post_response_url.append(response.headers["operation-location"])
+
+        # Get the data back from NLP API, convert code to conceptIDs
+        get_response = get_data_from_nlp(
+            url=url, headers=headers, post_response_url=post_response_url
+        )
+        codes = process_nlp_response(get_response)
+        # Look up standard and valid conceptIDs for concept codes
+        # Append conceptID to item in list, turn into a dictionary
+        codes_dict = concept_code_to_id(codes)
+
+        # Check each item in values and see whether NLP got a result
+        # If NLP finds something, save the result to ScanReportConcept
+        match = list(
+            filter(lambda item: item["pk"] == str(field.id), codes_dict))
+
+        for item in match:
+            scan_report_field = ScanReportField.objects.get(pk=item["pk"])
+            concept = Concept.objects.get(pk=item["conceptid"])
+
+            ScanReportConcept.objects.create(
+                nlp_entity=item["nlp_entity"],
+                nlp_entity_type=item["nlp_entity_type"],
+                nlp_confidence=item["nlp_confidence"],
+                nlp_vocabulary=item["nlp_vocab"],
+                nlp_concept_code=item["nlp_code"],
+                concept=concept,
+                content_object=scan_report_field,
+            )
+
+    else:
```

Comment:
These two if conditions are long, and could got into functions to improve readability.